src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/DocumentMapping/RemoteDocumentMappingService.cs

```diff
@@ -46,31 +47,18 @@ internal sealed class RemoteDocumentMappingService(
             return (generatedDocumentUri, generatedDocumentRange);
         }
 
-        var solution = originSnapshot.TextDocument.Project.Solution;
-        if (!solution.TryGetRazorDocument(razorDocumentUri, out var razorDocument))
-        {
-            return (generatedDocumentUri, generatedDocumentRange);
-        }
-
-        var razorDocumentSnapshot = _snapshotManager.GetSnapshot(razorDocument);
-
-        var razorCodeDocument = await razorDocumentSnapshot
-            .GetGeneratedOutputAsync(cancellationToken)
-            .ConfigureAwait(false);
-
-        if (razorCodeDocument is null)
-        {
-            return (generatedDocumentUri, generatedDocumentRange);
-        }
-
-        if (!razorCodeDocument.TryGetGeneratedDocument(generatedDocumentUri, _filePathService, out var generatedDocument))
-        {
-            return Assumed.Unreachable<(Uri, LinePositionSpan)>();
-        }
-
-        if (TryMapToHostDocumentRange(generatedDocument, generatedDocumentRange, MappingBehavior.Strict, out var mappedRange))
+        var project = originSnapshot.TextDocument.Project;
+        var solution = project.Solution;
+        if (await project.TryGetHintNameFromGeneratedDocumentUriAsync(generatedDocumentUri, cancellationToken).ConfigureAwait(false) is { } hintName &&
+            await project.TryGetSourceGeneratedDocumentFromHintNameAsync(hintName, cancellationToken).ConfigureAwait(false) is { } generatedDocument &&
+            await _snapshotManager.GetSnapshot(generatedDocument.Project).TryGetCodeDocumentFromGeneratedHintNameAsync(hintName, cancellationToken).ConfigureAwait(false) is { } razorCodeDocument &&
+            TryMapToHostDocumentRange(razorCodeDocument.GetCSharpDocument(), generatedDocumentRange, MappingBehavior.Strict, out var mappedRange))
         {
-            return (razorDocumentUri, mappedRange);
+            // TODO: Should we have a better way to do this? Store Uri in RazorCodeDocument?
```

Comment:
I would vote to _not_ store more things on RazorCodeDocument. Is it possible to store the Uri on the RemoteDocumentSnapshot?

Reply:
If its a strong vote for no, I'll remove the comment. We don't really have a RemoteDocumentSnapshot in this situation, but that actually might be a nice evolution. Instead of dealing with a TextDocument for the Razor file, and a SourceGeneratedDocument for the generated file, we could potentially lean on RemoteDocumentSnapshot as a convenient way to access both. I'll leave it for a future PR, but make a note for sure.